### PR TITLE
Fix how we update `CostModels`

### PR DESCRIPTION
**Pull request by lehins**
# Description

This is a follow up to #3953 

There are a couple of problems with CostModels in Conway:

* first one is that we retain invalid and unknown CostModels indefinitely. This PR clears them out as soon as a valid replacement has been added
* ToJSON/FromJSON did not account for unknown and invalid CostModels, thus information was lost during roundtrip and there was no way for a user to see this very useful information
* Constructor for `CostMododels` was exported thus it was very easy to invalidate invariant that CBOR and JSON instances rely on. Creation and modification of `CostModels` is done with smart constructors and through `Monoid` instance

Besides fixing CostModels this PR also adds more tests and organizes a little bit by removing duplicate functionality

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs [hunk 1]
```diff
@@ -286,18 +297,47 @@ decodeCostModelFailHard lang = do
...
+        P.CMUnknownParamError {} -> 0
+        P.CMInternalReadError {} -> 1
+        P.CMInternalWriteError {} -> 2
+        P.CMTooFewParamsError {} -> 3
+
+instance EncCBOR CostModelError where
+  encCBOR =
+    encode . \case
+      CostModelError (P.CMUnknownParamError err) ->
+        Sum (CostModelError . P.CMUnknownParamError) 1 !> To err
```
Comment (teodanciu) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs]: Why start at 1 and not at 0 ?

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus.hs [hunk 2]
```diff
@@ -30,6 +31,14 @@ import qualified PlutusLedgerApi.Test.V2.EvaluationContext as PV2
 import qualified PlutusLedgerApi.Test.V3.EvaluationContext as PV3
 import PlutusLedgerApi.V1 as PV1
 
+-- | Construct a test cost model where all paramters are set to the same value
```
Comment (teodanciu) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus.hs]: Small typo `parameters`

File: libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs [hunk 3]
```diff
@@ -37,7 +37,12 @@ module Cardano.Ledger.Plutus.CostModels (
   costModelToMap,
   costModelFromMap,
   decodeCostModelFailHard,
-  CostModels (..),
+  CostModels,
```
Comment (teodanciu) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs:40]: Maybe this will be easier to find if we move it up after `CostModel` export ?

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Arbitrary.hs [hunk 4]
```diff
@@ -253,23 +253,22 @@ genValidCostModel lang = do
   either (\err -> error $ "Corrupt cost model: " ++ show err) pure $
     mkCostModel lang newParamValues
 
-genValidCostModelPair :: Language -> Gen (Language, CostModel)
-genValidCostModelPair lang = (,) lang <$> genValidCostModel lang
+genValidCostModels :: Set.Set Language -> Gen CostModels
+genValidCostModels = fmap mkCostModels . sequence . Map.fromSet genValidCostModel
```
Comment (teodanciu) [inline eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Arbitrary.hs]: This is not strictly related to this change, but since we removed one useless function (genValidCostModelPair), maybe we could also remove its exact copy a few lines above, `genCostModel`, or at least reduce it's scope, since it's used only in [one place](https://github.com/IntersectMBO/cardano-ledger/pull/3954/commits/9d82ccdab94d04e52aa2bc7de35bc398b437624e#diff-334ba93db40dbbd431eb2831c9fe3226881e21c81945dc0dd3551ca9f215b62dL248), in the Arbitrary instance of Costmodel. 
Just because it's polluting a little the outline of the module that has become much clearer with this PR. And also it's kinda confusing to have `genValidCostModel` and `genCostModel`, whereas both generate valid ones, but in a different form.  But just an idea, it doesn't make much difference

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs [hunk 5]
```diff
@@ -9,30 +10,43 @@
...
-spec pxy = do
-  describe "CostModels CBOR deserialization" $ do
-    validCostModelProp @era pxy
+spec :: forall era. AlonzoEraPParams era => Spec
+spec = do
+  describe "CBOR deserialization" $ do
+    validCostModelProp @era Proxy
     invalidCostModelProp @era Proxy
     unknownCostModelProp @era Proxy
+  prop "applyPPUpdates" $ \valid validUpdate unknown unknownUpdate -> do
```
Comment (teodanciu) [inline eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs:33]: The formatting looks a bit strange here, don't know why fourmolu doesn't pick it up, but it looks like it should be aligned with the line above `unKnownCostModelProp @era Proxy`, no?

File: eras/alonzo/impl/CHANGELOG.md [hunk 6]
```diff
@@ -53,6 +53,7 @@
 
 ### `testlib`
 
+* Remove `Test.Cardano.Ledger.Alonzo.CostModel` in favor of `Test.Cardano.Ledger.Plutus`
```
Comment (teodanciu) [inline eras/alonzo/impl/CHANGELOG.md:56]: Maybe we should also mention the removal of `costModelParamsCount` from Arbitrary

File: libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs [hunk 7]
```diff
@@ -355,63 +360,77 @@ instance NFData CostModelError
...
-    Just lang ->
-      case mkCostModel lang cmIds of
-        Right cm -> CostModels (Map.insert lang cm validCMs) errs invalidCMs
-        Left e -> CostModels validCMs (addError lang e errs) updatedInvalidCMs
-    Nothing -> CostModels validCMs errs updatedInvalidCMs
+updateCostModels (CostModels oldValid oldErrors oldUnk) (CostModels modValid modErrors modUnk) =
+  CostModels
+    newValid
+    (newErrors Map.\\ newValid)
+    (Map.union modUnk oldUnk Map.\\ Map.mapKeys languageToWord8 newValid)
```
Comment (teodanciu) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs:422]: Nice, so this keeps the property that a Language is either in the valid or errors map that we have when using `mkCostModelsLenient`.   The resulting unknown costmodels  could still have keys that are in `errors`  (which is not possible in `mkCostModelsLenient`), but  don't think it matters.

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs (hunk 5 above)
Comment (lehins) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs:33]: No, this was intentional, since this property has nothing to do with serialization. It is just part of the `CostModels` `Spec`

File: libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs [hunk 8]
```diff
@@ -366,6 +367,12 @@ data CostModels = CostModels
   }
   deriving stock (Eq, Ord, Show, Generic)
 
+instance Semigroup CostModels where
+  (<>) = updateCostModels
```
Comment (teodanciu) [inline libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs:391]: Why not use it in [PParams.ppUpdateCostModels](https://github.com/IntersectMBO/cardano-ledger/blob/lehins/cost-models-fixes/eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs#L1192)   ?
`SJust costModelUpdate -> THKD $ curCostModel <> costModelUpdate`   

Any particular reason?

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs (hunk 5 above)
Comment (teodanciu) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs:33]: Oh yes, makes sense!  Missed the content of "describe"

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs [hunk 9]
```diff
@@ -9,30 +10,43 @@
...
+    let
+      validExpected = mkCostModels (costModelsValid validUpdate <> costModelsValid valid)
+      unknownExpected = unknownUpdate <> unknown
```
Comment (teodanciu) [inline eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs:36]: I suppose the chance of the generated `unknown` to be actually a valid one can be neglected

File: eras/alonzo/test-suite/test/Test/Cardano/Ledger/Alonzo/Golden.hs [hunk 10]
```diff
@@ -17,21 +17,24 @@ import Cardano.Ledger.Alonzo.PParams (
...
-  mkCostModel,
- )
 import Cardano.Ledger.Alonzo.TxBody (AlonzoTxOut (..), utxoEntrySize)
 import Cardano.Ledger.BaseTypes (StrictMaybe (..), boundRational)
 import Cardano.Ledger.Binary (decCBOR, decodeFullAnnotator)
 import Cardano.Ledger.Binary.Plain as Plain (serialize)
 import Cardano.Ledger.Block (Block (..))
 import Cardano.Ledger.Coin (Coin (..))
 import Cardano.Ledger.Mary.Value (valueFromList)
+import Cardano.Ledger.Plutus.CostModels (
```
Comment (teodanciu) [inline eras/alonzo/test-suite/test/Test/Cardano/Ledger/Alonzo/Golden.hs:27]: Shouldn't we update the dependency on cardano-ledger-core  in [cardano-ledger-alonzo-test.cabal ](https://github.com/IntersectMBO/cardano-ledger/blob/lehins/cost-models-fixes/eras/alonzo/test-suite/cardano-ledger-alonzo-test.cabal#L49) to 

` cardano-ledger-core:{cardano-ledger-core, testlib} >=1.10 && <1.11`  ?

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs (hunk 9 above)
Comment (lehins) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs:36]: On the contrary, it is desired. I actually found a bug in my implementation earlier thanks to the fact that we generate unknown updates that can be valid

File: libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs (hunk 3 above)
Comment (lehins) [reply libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs:40]: It'll actually be better if I'll organize functions and type relative to `CostModel` vs `CostModels`. So'll create a section for `CostModels` and will move this export to the top of that section.

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Arbitrary.hs (hunk 4 above)
Comment (lehins) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Arbitrary.hs]: Great catch. Done

File: eras/alonzo/impl/CHANGELOG.md (hunk 6 above)
Comment (lehins) [reply eras/alonzo/impl/CHANGELOG.md:56]: Done.

File: libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs (hunk 8 above)
Comment (lehins) [reply libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs:391]: No reason. It's not really worth a change though.

File: eras/alonzo/test-suite/test/Test/Cardano/Ledger/Alonzo/Golden.hs (hunk 10 above)
Comment (lehins) [reply eras/alonzo/test-suite/test/Test/Cardano/Ledger/Alonzo/Golden.hs:27]: Yes. Nice catch. Done.

File: libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs (hunk 7 above)
Comment (lehins) [reply libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs:422]: > The resulting unknown costmodels could still have keys that are in errors (which is not possible in mkCostModelsLenient), but don't think it matters.

That is actually necessary for rountripping. If we were to remove costmodels that caused errors from unknown then the errors field would be lost, since we do not serialize errors.

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus.hs (hunk 2 above)
Comment (lehins) [reply libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus.hs]: Fixed

**Review by teodanciu — APPROVED** [review]
Looks good to me

File: eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs (hunk 9 above)
Comment (teodanciu) [reply eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs:36]: I see, because we're flattening them after. Makes sense.

File: libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs (hunk 1 above)
Comment (lehins) [reply libs/cardano-ledger-core/src/Cardano/Ledger/Plutus/CostModels.hs]: Funny you say that. I wrote it initially starting with `0`, naturally. Then I looked at serialization of some predicate failure, which started at `1` and I thought to myself: "Oh, I guess we start at `1` for those decoders", so I switched it to start at `1` for "consistency". But now I can't even find that predicate failure :smile: 

I've switched to start at 0, thank you for bringing this up!
